File: remote-notify.pl
#!/usr/bin/env perl
use strict;
use HTTP::Request;
use Irssi;
use LWP::UserAgent;
use POSIX;
use URI::Escape;
use vars qw($VERSION %IRSSI);
$VERSION = '1.00';
%IRSSI = (
    authors     => 'Geoffrey Anderson',
    contact     => 'geoff@geoffreyanderson.net',
    name        => 'Remote Notify',
    description => 'This script allows you to ' .
                   'send notifications from irssi ' .
                   'to a notify-osd or growl (OS X) ' .
                   'through an SSH tunnel.',
    license     => 'GNU GPL v3',
);
my $URL_TEMPLATE = "http://localhost:%d?title=%s&message=%s";
my $NOTIFY_TITLE = "irssi mention";
my $UA = LWP::UserAgent->new;


sub remote_notify {
    # Grab the default values sent to this signal
    my ($dest, $text, $stripped) = @_;

    # get a reference to the active window in irssi
    my $window = Irssi::active_win();

    # grab the port from the settings
    my $port = Irssi::settings_get_int('remotenotify_port');

    # Check if the printed message is a hilight
    if ( ($dest->{level} & (MSGLEVEL_HILIGHT|MSGLEVEL_MSGS)) &&
        ($dest->{level} & MSGLEVEL_NOHILIGHT) == 0)
    {
        # Strip all the funky irssi color coding out of the text
        $text = Irssi::strip_codes($text);

        # if the hilighted message is in a channel, prepend the channel name to
        # the message text
        if ($dest->{level} & MSGLEVEL_PUBLIC) {
            $text = $dest->{target}.": ".$text;
        }

        # Prepend a timestamp to the message text
        $text = strftime(Irssi::settings_get_str('timestamp_format')." ", localtime).$text;

        # Create the HTTP request using the template and parameters collected
        # above
        my $request = HTTP::Request->new(
            GET => sprintf($URL_TEMPLATE, $port, uri_escape($NOTIFY_TITLE), uri_escape($text))
        );

        # Get the response. Print to the irssi window if an error was detected.
        my $response = $UA->request($request);
        if (!$response->is_success) {
            $window->print("Failed to send data to remote notifier.", MSGLEVEL_NEVER) if ($window);
        }
    }
}

Irssi::settings_add_int('remotenotify', 'remotenotify_port', 4235);
Irssi::signal_add("print text", "remote_notify");

File: remote-notify.py
#!/usr/bin/env python
# Copyright (c) 2012, Dave Eddy <dave@daveeddy.com>
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#	 * Redistributions of source code must retain the above copyright
#	   notice, this list of conditions and the following disclaimer.
#	 * Redistributions in binary form must reproduce the above copyright
#	   notice, this list of conditions and the following disclaimer in the
#	   documentation and/or other materials provided with the distribution.
#	 * Neither the name of the project nor the
#	   names of its contributors may be used to endorse or promote products
#	   derived from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# Allow remote notifications for weechat in growl/osd-notify
#
# :: To use ::
# On the remote machine running weechat:
#	Enable the plugin
#	/python load remote-notify.py
# On the machine you are connecting from:
#	Start the service.
#	~$ ./remote-notify.py
#	* This starts the web service listening on localhost
#	Then, ssh to the host running weechat and forward the port
#	~$ ssh -R 4235:localhost:4235 user@host
#

__program__     = 'remote-notify'
__author__      = 'David Eddy'
__credits__     = ['Michael Zeller']
__license__     = 'BSD 3-Clause'
__version__     = '0.1'
__maintainer__  = 'David Eddy'
__email__       = 'dave@daveeddy.com'
__status__      = 'Release'
__description__ = 'Send remote notifications to growl/osd-notify'

from os import uname

in_weechat = True
try:
	import weechat
except ImportError:
	in_weechat = False

# Default port and host to run the server
DEFAULT_PORT = 4235
DEFAULT_HOST = '127.0.0.1'

# Uname of the machine
UNAME = uname()[0]

# Commands to use to make notifications based on uname
NOTIFY_COMMANDS={
	'Darwin'  : 'growlnotify',
	'default' : 'notify-send',
}

def r_callback(data, signal, signal_data):
	"""
	Called from weechat

	Take the data passed from weechat and
	attempt to send the data to the server
	"""
	# Split the signal data to get the sender and the message
	msg_sender, message = signal_data.split('\t', 1)

	# Construct the GET paramaters to use for the notification
	data = {
		'title'   : 'Mentioned by %s' % (msg_sender),
		'message' : message,
	}
	# Make the request URL
	url = 'http://localhost:%d?%s' % (DEFAULT_PORT, urllib.urlencode(data))
	try:
		# Try to make the request
		output = urllib2.urlopen(url)
	except URLError:
		weechat.prnt('', 'Failed to send remote notification')
	# Do nothing with output
	return weechat.WEECHAT_RC_OK

def notify(title, message):
	"""
	Send a notification on a system
	"""
	if UNAME == 'Darwin':
		# Construct the mac command
		cmd = [NOTIFY_COMMANDS[UNAME], '-m', message, '-t', title]
	else:
		# Construct the default command
		cmd = [NOTIFY_COMMANDS[UNAME], title, message]
	try:
		# Call the subprocess
		p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
	except OSError:
		return None
	# Return the Popen object
	return p

if __name__ == '__main__':
	if in_weechat:
		# Running inside weechat
		if weechat.register(__program__, __author__, __version__, __license__, __description__, '', ''):
			import urllib, urllib2
			weechat.hook_signal('weechat_highlight', 'r_callback', '')
	else:
		# Running from the command line
		import BaseHTTPServer
		import urlparse
		import cgi
		import subprocess

		class MyHandler(BaseHTTPServer.BaseHTTPRequestHandler):
			"""
			Simple HTTP class to listen for GET requests
			"""
			def do_GET(s):
				"""Respond to a GET request."""
				try:
					# try to parse the request
					parsed_path = urlparse.urlparse(s.path)
					q = cgi.parse_qs(parsed_path.query)
					title = q['title'][0]
					message = q['message'][0]
					# Send the notification
					p = notify(title, message)
					# Check for errors
					if not p:
						send = 'Command %s not found' % (NOTIFY_COMMANDS[UNAME])
					else:
						err = p.stderr.read()
						if err:
							send = err
						else:
							send = 'Message Sent Successfully'
				except:
					send = 'Error: Error parsing message'
				s.send_response(200)
				s.send_header("Content-type", "text/html")
				s.end_headers()
				s.wfile.write('<status>%s</status>' % (send))
				print send

		# Create the httpd service
		server_class = BaseHTTPServer.HTTPServer
		httpd = server_class((DEFAULT_HOST, DEFAULT_PORT), MyHandler)
		try:
			# Start the service
			print 'Listening on %s:%d' % (DEFAULT_HOST, DEFAULT_PORT)
			httpd.serve_forever()
		except KeyboardInterrupt:
			httpd.socket.close()
